File: backend/rust/src/server/control.rs
use std::collections::HashSet;
use std::net::SocketAddr;
use std::sync::atomic::Ordering::Relaxed;
use std::sync::Arc;
use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};

use axum::extract::ws::{CloseFrame, Message, WebSocket};
use futures_util::stream::{SplitSink, SplitStream};
use futures_util::{SinkExt, StreamExt};
use tokio::sync::mpsc;
use tracing::{debug, info};

use super::session::{random_bytes, KeySlot, PeerState, Session, TokenBucket, UdpState};
use super::Server;
use crate::auth::{self, AuthError, Identity};
use crate::protocol::control::*;
use crate::protocol::udp::{MAJOR, MINOR};
use crate::routing::{scope_key, Peer};

const HELLO_TIMEOUT: Duration = Duration::from_secs(10);
const AUTH_TIMEOUT: Duration = Duration::from_secs(30);

type Sink = SplitSink<WebSocket, Message>;
type Stream = SplitStream<WebSocket>;

fn unix_now() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or(0)
}

fn unix_ms() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0)
}

enum ReadErr {
    Timeout,
    Closed,
    Binary,
}

async fn read_text(stream: &mut Stream, timeout: Duration) -> Result<String, ReadErr> {
    let deadline = tokio::time::Instant::now() + timeout;
    loop {
        match tokio::time::timeout_at(deadline, stream.next()).await {
            Err(_) => return Err(ReadErr::Timeout),
            Ok(None) | Ok(Some(Err(_))) => return Err(ReadErr::Closed),
            Ok(Some(Ok(Message::Text(t)))) => return Ok(t.to_string()),
            Ok(Some(Ok(Message::Binary(_)))) => return Err(ReadErr::Binary),
            Ok(Some(Ok(Message::Close(_)))) => return Err(ReadErr::Closed),
            Ok(Some(Ok(_))) => continue, // ping/pong handled by the library
        }
    }
}

async fn send_text(sink: &mut Sink, s: String) -> bool {
    tokio::time::timeout(Duration::from_secs(5), sink.send(Message::Text(s.into())))
        .await
        .is_ok_and(|r| r.is_ok())
}

async fn fail(sink: &mut Sink, code_: &'static str, msg: &str) {
    let _ = send_text(sink, error_frame(code_, msg, true)).await;
    let status = if code_ == code::INTERNAL { 1011 } else { 1008 };
    let _ = tokio::time::timeout(
        Duration::from_secs(2),
        sink.send(Message::Close(Some(CloseFrame {
            code: status,
            reason: code_.into(),
        }))),
    )
    .await;
}

fn hex_u64(v: u64) -> String {
    hex::encode(v.to_be_bytes())
}

pub(crate) async fn handle_socket(srv: Arc<Server>, socket: WebSocket, ip: String, host: String) {
    let (mut sink, mut stream) = socket.split();
    let tag = srv.ip_tag(&ip);

    // --- hello ---
    let frame = match read_text(&mut stream, HELLO_TIMEOUT).await {
        Ok(f) => f,
        Err(_) => return fail(&mut sink, code::BAD_MESSAGE, "expected hello").await,
    };
    let hello = match parse_control(frame.as_bytes()) {
        Ok(ClientMsg::Hello(h)) => h,
        Ok(_) => return fail(&mut sink, code::BAD_MESSAGE, "first message must be hello").await,
        Err(e) => return fail(&mut sink, e.code, e.message).await,
    };
    let challenge = hex::encode(random_bytes::<16>());
    let ok = HelloOk {
        t: "hello_ok",
        protocol: Version {
            major: MAJOR as u32,
            minor: MINOR as u32,
        },
        server: ServerInfo {
            name: "mcvoice-backend",
            version: super::VERSION,
            implementation: "rust",
        },
        capabilities: SERVER_CAPABILITIES,
        auth: AuthInfo {
            methods: vec![srv.cfg.auth_mode.clone()],
            challenge: challenge.clone(),
        },
    };
    if !send_text(&mut sink, serde_json::to_string(&ok).unwrap()).await {
        return;
    }
    let challenge_at = Instant::now();

    // --- auth / resume ---
    let frame = match read_text(&mut stream, AUTH_TIMEOUT).await {
        Ok(f) => f,
        Err(_) => return fail(&mut sink, code::AUTH_REQUIRED, "expected auth").await,
    };
    let auth_failed = |srv: &Server| {
        srv.metrics.auth_failures.fetch_add(1, Relaxed);
        srv.auth_fail_limiter.allow(&ip, Instant::now());
    };
    let ident: Identity = match parse_control(frame.as_bytes()) {
        Err(e) => return fail(&mut sink, e.code, e.message).await,
        Ok(ClientMsg::Auth(a)) => {
            let method = a.method.unwrap();
            let username = a.username.unwrap();
            if method != srv.cfg.auth_mode {
                return fail(&mut sink, code::UNSUPPORTED_AUTH_METHOD, "auth method not enabled").await;
            }
            if challenge_at.elapsed() > Duration::from_secs(60) {
                return fail(&mut sink, code::AUTH_FAILED, "challenge expired").await;
            }
            if method == "mojang" {
                match srv.mojang.verify(&username, &challenge).await {
                    Ok(id) => id,
                    Err(e) => {
                        auth_failed(&srv);
                        info!(category = "control", ip = %tag, error = ?e, "authentication failed");
                        return fail(&mut sink, code::AUTH_FAILED, "could not verify Minecraft session").await;
                    }
                }
            } else {
                Identity {
                    uuid: a.uuid.unwrap(),
                    username,
                }
            }
        }
        Ok(ClientMsg::Resume(r)) => match auth::verify_resume(&srv.cfg.jwt_signing_secret, &r.resume_token.unwrap(), unix_now()) {
            Some(id) => id,
            None => {
                auth_failed(&srv);
                return fail(&mut sink, code::SESSION_EXPIRED, "resume token invalid or expired").await;
            }
        },
        Ok(_) => return fail(&mut sink, code::AUTH_REQUIRED, "authenticate first").await,
    };
    let _ = AuthError::Rejected; // variants used by verify()
    if srv.is_banned(&ident.uuid) {
        return fail(&mut sink, code::BANNED, "banned").await;
    }

    let now = Instant::now();
    let ttl = srv.key_ttl();
    let (tx, mut rx) = mpsc::channel::<String>(128);
    let uuid_b = uuid_bytes(&ident.uuid).expect("validated uuid");
    let udp = UdpState {
        cur: KeySlot::new(0, ttl, now),
        prev: None,
        send_with_new: false,
        addr: None,
        udp_ok_sent: false,
        voice_bucket: TokenBucket::new(srv.cfg.rate_voice_per_sec, srv.cfg.rate_voice_burst),
        jitter: 0.0,
        last_transit: None,
        last_voice: None,
    };
    let sess = Arc::new(Session::new(ident.clone(), uuid_b, tx, udp));
    let admin_muted = srv.is_admin_muted(&ident.uuid);
    let ps = PeerState {
        peer: Peer {
            uuid: ident.uuid.clone(),
            authenticated: true,
            muted: admin_muted,
            ..Default::default()
        },
        network_id: String::new(),
        world_id: String::new(),
        attested: String::new(),
        peers_rev: 0,
        presence: HashSet::new(),
        self_muted: false,
        admin_muted,
        has_scope: false,
    };
    if !srv.register(sess.clone(), ps) {
        return fail(&mut sink, code::SERVER_FULL, "server full").await;
    }
    info!(category = "control", ip = %tag, session = %sess.id, player = %ident.uuid,
          client = %hello.client.as_ref().map(|c| c.version.as_str()).unwrap_or(""),
          minecraft = %hello.client.as_ref().map(|c| c.minecraft.as_str()).unwrap_or(""), "session started");

    let (key_id, key_b64) = {
        let u = sess.udp.lock().unwrap();
        (u.cur.id, u.cur.b64())
    };
    let port = if srv.cfg.public_voice_port != 0 {
        srv.cfg.public_voice_port
    } else {
        srv.udp.get().and_then(|s| s.local_addr().ok()).map(|a| a.port()).unwrap_or(0)
    };
    let msg = SessionMsg {
        t: "session",
        session_id: sess.id.clone(),
        player_uuid: ident.uuid.clone(),
        username: ident.username.clone(),
        resume_token: auth::issue_resume(&srv.cfg.jwt_signing_secret, &ident, srv.cfg.resume_token_ttl, unix_now()),
        voice: VoiceInfo {
            host: if srv.cfg.public_hostname.is_empty() {
                host
            } else {
                srv.cfg.public_hostname.clone()
            },
            port,
            connection_id: hex_u64(sess.conn_id),
            key_id,
            key: key_b64,
            key_expires_in: ttl.as_secs(),
        },
        config: SessionConfig {
            normal_range: srv.cfg.normal_range,
            whisper_range: srv.cfg.whisper_range,
            max_range: srv.cfg.max_range,
            codec: "opus",
            sample_rate: 48000,
            frame_ms: 20,
            heartbeat_interval: 5,
            position_hz_max: 10,
        },
    };
    if !send_text(&mut sink, serde_json::to_string(&msg).unwrap()).await {
        srv.unregister(&sess);
        return;
    }

    // writer task: drains the session queue, then sends the close frame
    let wsess = sess.clone();
    let writer = tokio::spawn(async move {
        let mut closed = wsess.closed_rx();
        loop {
            tokio::select! {
                m = rx.recv() => match m {
                    Some(m) => if !send_text(&mut sink, m).await { wsess.close("write_failed"); return; },
                    None => return,
                },
                _ = closed.changed() => {
                    while let Ok(m) = rx.try_recv() {
                        if !send_text(&mut sink, m).await { return; }
                    }
                    let code_ = wsess.close_code();
                    let status = if code_ == "bye" || code_ == "timeout" || code_ == "client_closed" { 1000 } else { 1008 };
                    let _ = tokio::time::timeout(Duration::from_secs(2),
                        sink.send(Message::Close(Some(CloseFrame { code: status, reason: code_.into() })))).await;
                    return;
                }
            }
        }
    });

    // key rotation task
    let rsrv = srv.clone();
    let rsess = sess.clone();
    let rotation = tokio::spawn(async move {
        let period = Duration::from_secs(rsrv.cfg.key_rotation_secs);
        let mut t = tokio::time::interval_at(tokio::time::Instant::now() + period, period);
        let mut closed = rsess.closed_rx();
        loop {
            tokio::select! {
                _ = t.tick() => rsrv.rotate_session(&rsess),
                _ = closed.changed() => return,
            }
        }
    });

    let timeout = Duration::from_secs(srv.cfg.session_timeout_secs);
    let mut ctl_bucket = TokenBucket::new(srv.cfg.rate_control_per_sec, srv.cfg.rate_control_burst);
    let mut last_pos: Option<Instant> = None;
    let (mut strikes, mut strike_window) = (0u32, Instant::now());
    let mut closed = sess.closed_rx();
    let reason: &'static str = loop {
        let frame = tokio::select! {
            r = read_text(&mut stream, timeout) => r,
            _ = closed.changed() => break "closed",
        };
        let frame = match frame {
            Ok(f) => f,
            Err(ReadErr::Timeout) => {
                sess.close("timeout");
                break "timeout";
            }
            Err(ReadErr::Closed) => {
                sess.close("client_closed");
                break "client_closed";
            }
            Err(ReadErr::Binary) => {
                sess.send(error_frame(code::BAD_MESSAGE, "binary frames are not allowed", true));
                sess.close(code::BAD_MESSAGE);
                break "binary";
            }
        };
        srv.metrics.control_messages.fetch_add(1, Relaxed);
        let now = Instant::now();
        sess.last_seen.store(unix_ms(), Relaxed);
        if now.duration_since(strike_window) > Duration::from_secs(10) {
            strikes = 0;
            strike_window = now;
        }
        if !ctl_bucket.allow(now) {
            strikes += 1;
            if strikes > 200 {
                sess.send(error_frame(code::RATE_LIMITED, "sustained control flood", true));
                sess.close(code::RATE_LIMITED);
                break "rate_limited";
            }
            sess.send(error_frame(code::RATE_LIMITED, "slow down", false));
            continue;
        }
        match parse_control(frame.as_bytes()) {
            Err(e) if e.code == code::UNKNOWN_MESSAGE => sess.send(error_frame(e.code, e.message, false)),
            Err(e) => {
                sess.send(error_frame(e.code, e.message, true));
                sess.close(e.code);
                break "bad_message";
            }
            Ok(ClientMsg::Bye) => {
                sess.close("bye");
                break "bye";
            }
            Ok(m) => srv.dispatch(&sess, m, now, &mut last_pos),
        }
    };
    srv.unregister(&sess);
    sess.close(reason);
    rotation.abort();
    let _ = tokio::time::timeout(Duration::from_secs(3), writer).await;
    info!(category = "control", session = %sess.id, reason = %if reason == "closed" { sess.close_code() } else { reason }, "session ended");
}

impl Server {
    fn dispatch(&self, sess: &Arc<Session>, msg: ClientMsg, now: Instant, last_pos: &mut Option<Instant>) {
        match msg {
            ClientMsg::Hello(_) | ClientMsg::Auth(_) | ClientMsg::Resume(_) => {
                sess.send(error_frame(code::BAD_MESSAGE, "already authenticated", false))
            }
            ClientMsg::Scope(s) => {
                let epoch = s.epoch.unwrap();
                let in_world = s.in_world.unwrap();
                let mut hub = self.hub.write().unwrap();
                let Some(ps) = hub.peers.get(&sess.conn_id) else { return };
                if ps.has_scope && epoch <= ps.peer.epoch {
                    drop(hub);
                    sess.send(error_frame(code::STALE_EPOCH, "epoch must increase", false));
                    return;
                }
                hub.leave_bucket(sess.conn_id);
                *last_pos = None; // the first position of a new epoch is always accepted
                let attested = if in_world {
                    s.attestation.as_deref().and_then(|a| {
                        let r = auth::verify_attestation(&self.cfg.attestation_keys, a, &sess.ident.uuid, unix_now());
                        if r.is_none() {
                            debug!(category = "control", "ignored invalid scope attestation");
                        }
                        r
                    })
                } else {
                    None
                };
                let ps = hub.peers.get_mut(&sess.conn_id).unwrap();
                ps.has_scope = true;
                ps.peer.epoch = epoch;
                ps.peer.in_world = in_world;
                ps.peer.pos = None;
                ps.peer.visible.clear();
                ps.peers_rev = 0;
                ps.presence.clear();
                ps.attested = attested.unwrap_or_default();
                if in_world {
                    ps.network_id = s.network_id.unwrap();
                    ps.world_id = s.world_id.unwrap();
                    let key = scope_key(&ps.network_id, &ps.attested, &ps.world_id);
                    hub.join_bucket(sess.conn_id, key);
                }
            }
            ClientMsg::Pos(p) => {
                if last_pos.is_some_and(|t| now.duration_since(t) < Duration::from_millis(90)) {
                    return; // above position_hz_max: dropped silently
                }
                *last_pos = Some(now);
                let (e, x, y, z) = (p.epoch.unwrap(), p.x.unwrap(), p.y.unwrap(), p.z.unwrap());
                let now_ms = self.now_ms();
                let mut hub = self.hub.write().unwrap();
                if let Some(ps) = hub.peers.get_mut(&sess.conn_id) {
                    if e == ps.peer.epoch && ps.peer.in_world {
                        ps.peer.pos = Some([x, y, z]);
                        ps.peer.pos_at_ms = now_ms;
                    }
                }
                drop(hub);
                if self.cfg.log_positions {
                    debug!(category = "control", epoch = e, x, y, z, "position");
                }
            }
            ClientMsg::Peers(p) => {
                let mut hub = self.hub.write().unwrap();
                if let Some(ps) = hub.peers.get_mut(&sess.conn_id) {
                    if p.epoch.unwrap() == ps.peer.epoch {
                        ps.peer.visible = p.full.unwrap().into_iter().filter(|u| *u != sess.ident.uuid).collect();
                        ps.peers_rev = p.rev.unwrap();
                    }
                }
            }
            ClientMsg::PeersDelta(d) => {
                let mut hub = self.hub.write().unwrap();
                let Some(ps) = hub.peers.get_mut(&sess.conn_id) else { return };
                if d.epoch.unwrap() != ps.peer.epoch {
                    return;
                }
                let (base, rev) = (d.base.unwrap(), d.rev.unwrap());
                if base != ps.peers_rev || rev <= base {
                    let epoch = ps.peer.epoch;
                    drop(hub);
                    sess.send(serde_json::to_string(&EpochMsg { t: "peers_resync", epoch }).unwrap());
                    return;
                }
                for u in d.remove.unwrap_or_default() {
                    ps.peer.visible.remove(&u);
                }
                for u in d.add.unwrap_or_default() {
                    if u != sess.ident.uuid && ps.peer.visible.len() < MAX_PEER_LIST {
                        ps.peer.visible.insert(u);
                    }
                }
                ps.peers_rev = rev;
            }
            ClientMsg::State(s) => {
                let mut hub = self.hub.write().unwrap();
                if let Some(ps) = hub.peers.get_mut(&sess.conn_id) {
                    ps.self_muted = s.muted.unwrap();
                    ps.peer.muted = ps.self_muted || ps.admin_muted;
                    ps.peer.deafened = s.deafened.unwrap();
                }
            }
            ClientMsg::Ping(p) => sess.send(
                serde_json::to_string(&PongMsg {
                    t: "pong",
                    nonce: p.nonce.unwrap(),
                    server_time: unix_ms(),
                })
                .unwrap(),
            ),
            ClientMsg::Bye => {}
        }
    }
}

/// Voice host announced to clients when PUBLIC_HOSTNAME is unset.
pub(crate) fn host_only(host_header: &str) -> String {
    if let Ok(sa) = host_header.parse::<SocketAddr>() {
        return sa.ip().to_string();
    }
    match host_header.rsplit_once(':') {
        Some((h, p)) if p.chars().all(|c| c.is_ascii_digit()) && !h.contains(':') => h.to_string(),
        _ => host_header.trim_start_matches('[').trim_end_matches(']').to_string(),
    }
}

File: backend/go/internal/server/control.go
package server

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"net"
	"net/http"
	"strings"
	"time"

	"github.com/coder/websocket"

	"github.com/RavoxX/MCVoice/backend/go/internal/auth"
	"github.com/RavoxX/MCVoice/backend/go/internal/protocol"
	"github.com/RavoxX/MCVoice/backend/go/internal/routing"
)

const (
	helloTimeout = 10 * time.Second
	authTimeout  = 30 * time.Second
)

func (s *Server) clientIP(r *http.Request) string {
	if s.cfg.TrustProxyHeaders {
		if xff := r.Header.Get("X-Forwarded-For"); xff != "" {
			first, _, _ := strings.Cut(xff, ",")
			return strings.TrimSpace(first)
		}
	}
	host, _, err := net.SplitHostPort(r.RemoteAddr)
	if err != nil {
		return r.RemoteAddr
	}
	return host
}

func (s *Server) voiceHost(r *http.Request) string {
	if s.cfg.PublicHostname != "" {
		return s.cfg.PublicHostname
	}
	host, _, err := net.SplitHostPort(r.Host)
	if err != nil {
		return r.Host
	}
	return host
}

type ctlConn struct {
	ws  *websocket.Conn
	ctx context.Context
}

func (c *ctlConn) write(b []byte) error {
	ctx, cancel := context.WithTimeout(c.ctx, 5*time.Second)
	defer cancel()
	return c.ws.Write(ctx, websocket.MessageText, b)
}

func (c *ctlConn) read(timeout time.Duration) ([]byte, error) {
	ctx, cancel := context.WithTimeout(c.ctx, timeout)
	defer cancel()
	t, b, err := c.ws.Read(ctx)
	if err != nil {
		return nil, err
	}
	if t != websocket.MessageText {
		return nil, errBinary
	}
	return b, nil
}

var errBinary = errors.New("binary frame")

func (c *ctlConn) fail(code, msg string) {
	_ = c.write(errorFrame(code, msg, true))
	status := websocket.StatusPolicyViolation
	if code == protocol.CodeInternal {
		status = websocket.StatusInternalError
	}
	_ = c.ws.Close(status, code)
}

// HandleControl serves GET /v1/control.
func (s *Server) HandleControl(w http.ResponseWriter, r *http.Request) {
	ip := s.clientIP(r)
	now := s.now()
	if s.shutdown.Load() {
		http.Error(w, "shutting down", http.StatusServiceUnavailable)
		return
	}
	if !s.connectLimiter.allow(ip, now) || s.authFailLimiter.blocked(ip, now) {
		http.Error(w, "rate limited", http.StatusTooManyRequests)
		return
	}
	ws, err := websocket.Accept(w, r, &websocket.AcceptOptions{CompressionMode: websocket.CompressionDisabled})
	if err != nil {
		return
	}
	ws.SetReadLimit(protocol.MaxControlFrame + 1024)
	ctx, cancel := context.WithCancel(r.Context())
	defer cancel()
	c := &ctlConn{ws: ws, ctx: ctx}
	log := s.log.With("category", "control", "ip", s.ipTag(ip))

	// --- hello ---
	frame, err := c.read(helloTimeout)
	if err != nil {
		c.fail(protocol.CodeBadMessage, "expected hello")
		return
	}
	msg, cerr := protocol.ParseControl(frame)
	if cerr != nil {
		c.fail(cerr.Code, cerr.Message)
		return
	}
	hello, ok := msg.(*protocol.Hello)
	if !ok {
		c.fail(protocol.CodeBadMessage, "first message must be hello")
		return
	}
	challenge := hex.EncodeToString(randomBytes(16))
	methods := []string{s.cfg.AuthMode}
	if err := c.write(marshal(protocol.HelloOK{Type: "hello_ok", Protocol: protocol.Version{Major: protocol.Major, Minor: protocol.Minor},
		Server:       protocol.ServerInfo{Name: "mcvoice-backend", Version: Version, Implementation: "go"},
		Capabilities: protocol.ServerCapabilities, Auth: protocol.AuthInfo{Methods: methods, Challenge: challenge}})); err != nil {
		return
	}
	challengeAt := s.now()
	_ = hello

	// --- auth / resume ---
	frame, err = c.read(authTimeout)
	if err != nil {
		c.fail(protocol.CodeAuthRequired, "expected auth")
		return
	}
	msg, cerr = protocol.ParseControl(frame)
	if cerr != nil {
		c.fail(cerr.Code, cerr.Message)
		return
	}
	var ident auth.Identity
	switch m := msg.(type) {
	case *protocol.Auth:
		if *m.Method != s.cfg.AuthMode {
			c.fail(protocol.CodeUnsupportedAuthMethod, "auth method not enabled")
			return
		}
		if s.now().Sub(challengeAt) > 60*time.Second {
			c.fail(protocol.CodeAuthFailed, "challenge expired")
			return
		}
		switch *m.Method {
		case "mojang":
			actx, acancel := context.WithTimeout(ctx, 10*time.Second)
			ident, err = s.mojang.Verify(actx, *m.Username, challenge)
			acancel()
		case "offline":
			ident, err = auth.Identity{UUID: *m.UUID, Username: *m.Username}, nil
		}
	case *protocol.Resume:
		ident, err = auth.VerifyResume(s.cfg.JWTSigningSecret, *m.ResumeToken, s.now())
		if err != nil {
			s.metrics.AuthFailures.Inc()
			s.authFailLimiter.allow(ip, s.now())
			c.fail(protocol.CodeSessionExpired, "resume token invalid or expired")
			return
		}
	default:
		c.fail(protocol.CodeAuthRequired, "authenticate first")
		return
	}
	if err != nil {
		s.metrics.AuthFailures.Inc()
		s.authFailLimiter.allow(ip, s.now())
		log.Info("authentication failed", "err", err)
		c.fail(protocol.CodeAuthFailed, "could not verify Minecraft session")
		return
	}
	if s.isBanned(ident.UUID) {
		c.fail(protocol.CodeBanned, "banned")
		return
	}

	now = s.now()
	ttl := time.Duration(s.cfg.KeyRotationSecs) * time.Second * 2
	sess := &Session{id: newSessionID(), connID: newConnID(), ident: ident,
		out: make(chan []byte, 128), done: make(chan struct{}),
		presence:    map[string]struct{}{},
		voiceBucket: newBucket(s.cfg.RateVoicePerSec, s.cfg.RateVoiceBurst),
		ctlBucket:   newBucket(s.cfg.RateControlPerSec, s.cfg.RateControlBurst)}
	sess.uuidB, _ = protocol.UUIDBytes(ident.UUID)
	sess.cur = newKeySlot(0, ttl, now)
	sess.adminMute = s.isAdminMuted(ident.UUID)
	sess.peer = routing.Peer{UUID: ident.UUID, Authenticated: true, Visible: map[string]struct{}{}, Muted: sess.adminMute}
	sess.lastSeen.Store(now.UnixMilli())
	if !s.register(sess) {
		c.fail(protocol.CodeServerFull, "server full")
		return
	}
	defer s.unregister(sess)
	log = log.With("session", sess.id)
	log.Info("session started", "player", ident.UUID, "client", hello.Client.Version, "minecraft", hello.Client.Minecraft, "loader", hello.Client.Loader)

	if err := c.write(marshal(protocol.SessionMsg{Type: "session", SessionID: sess.id, PlayerUUID: ident.UUID, Username: ident.Username,
		ResumeToken: auth.IssueResume(s.cfg.JWTSigningSecret, ident, time.Duration(s.cfg.ResumeTokenTTL)*time.Second, now),
		Voice: protocol.VoiceInfo{Host: s.voiceHost(r), Port: s.voicePort(), ConnectionID: hexU64(sess.connID),
			KeyID: int(sess.cur.id), Key: sess.cur.b64(), KeyExpiresIn: int(ttl.Seconds())},
		Config: protocol.SessionConfig{NormalRange: s.cfg.NormalRange, WhisperRange: s.cfg.WhisperRange, MaxRange: s.cfg.MaxRange,
			Codec: "opus", SampleRate: 48000, FrameMs: 20, HeartbeatInterval: 5, PositionHzMax: 10}})); err != nil {
		return
	}

	// writer
	go func() {
		for {
			select {
			case <-sess.done:
				cancel()
				return
			case <-ctx.Done():
				return
			case b := <-sess.out:
				if err := c.write(b); err != nil {
					cancel()
					return
				}
			}
		}
	}()

	// key rotation
	go func() {
		t := time.NewTicker(time.Duration(s.cfg.KeyRotationSecs) * time.Second)
		defer t.Stop()
		for {
			select {
			case <-ctx.Done():
				return
			case <-sess.done:
				return
			case <-t.C:
				sess.umu.Lock()
				k := sess.rotate(ttl, s.now())
				sess.umu.Unlock()
				sess.send(marshal(protocol.KeyMsg{Type: "key", KeyID: int(k.id), Key: k.b64(), KeyExpiresIn: int(ttl.Seconds())}))
			}
		}
	}()

	timeout := time.Duration(s.cfg.SessionTimeoutSec) * time.Second
	rateStrikes, strikeWindow := 0, s.now()
	for {
		frame, err := c.read(timeout)
		if err != nil {
			select {
			case <-sess.done:
				_ = c.ws.Close(websocket.StatusPolicyViolation, sess.closeCode)
			default:
				if errors.Is(err, errBinary) {
					c.fail(protocol.CodeBadMessage, "binary frames are not allowed")
				} else {
					_ = c.ws.Close(websocket.StatusNormalClosure, "timeout")
				}
			}
			log.Info("session ended", "reason", closeReason(err, sess))
			return
		}
		s.metrics.ControlMessages.Inc()
		now := s.now()
		sess.lastSeen.Store(now.UnixMilli())
		if now.Sub(strikeWindow) > 10*time.Second {
			rateStrikes, strikeWindow = 0, now
		}
		if !sess.ctlBucket.allow(now) {
			rateStrikes++
			if rateStrikes > 200 {
				c.fail(protocol.CodeRateLimited, "sustained control flood")
				return
			}
			sess.send(errorFrame(protocol.CodeRateLimited, "slow down", false))
			continue
		}
		msg, cerr := protocol.ParseControl(frame)
		if cerr != nil {
			if cerr.Code == protocol.CodeUnknownMessage {
				sess.send(errorFrame(cerr.Code, cerr.Message, false))
				continue
			}
			c.fail(cerr.Code, cerr.Message)
			return
		}
		if done := s.dispatch(sess, msg, now, log); done {
			_ = c.ws.Close(websocket.StatusNormalClosure, "bye")
			return
		}
	}
}

func closeReason(err error, sess *Session) string {
	select {
	case <-sess.done:
		return sess.closeCode
	default:
	}
	if websocket.CloseStatus(err) != -1 {
		return "client_closed"
	}
	return "timeout_or_network"
}

func (s *Server) voicePort() int {
	if s.cfg.PublicVoicePort != 0 {
		return s.cfg.PublicVoicePort
	}
	return s.UDPAddr().Port
}

func hexU64(v uint64) string {
	var b [8]byte
	for i := 7; i >= 0; i-- {
		b[i] = byte(v)
		v >>= 8
	}
	return hex.EncodeToString(b[:])
}

// dispatch applies one authenticated control message. Returns true on bye.
func (s *Server) dispatch(sess *Session, msg any, now time.Time, log interface {
	Debug(string, ...any)
}) bool {
	switch m := msg.(type) {
	case *protocol.Hello, *protocol.Auth, *protocol.Resume:
		sess.send(errorFrame(protocol.CodeBadMessage, "already authenticated", false))
	case *protocol.Scope:
		s.mu.Lock()
		if sess.hasScope && *m.Epoch <= sess.peer.Epoch {
			s.mu.Unlock()
			sess.send(errorFrame(protocol.CodeStaleEpoch, "epoch must increase", false))
			return false
		}
		sess.hasScope = true
		sess.lastPos = time.Time{} // the first position of a new epoch is always accepted
		sess.peer.Epoch = *m.Epoch
		sess.peer.InWorld = *m.InWorld
		sess.peer.HasPos = false
		sess.peer.Visible = map[string]struct{}{}
		sess.peersRev = 0
		sess.presence = map[string]struct{}{}
		sess.peer.WorldID, sess.peer.Attested = "", ""
		if *m.InWorld {
			// network_id is validated by the parser but never used for routing (spec 6.3)
			sess.peer.WorldID = *m.WorldID
			if m.Attestation != nil {
				if a, ok := auth.VerifyAttestation(s.cfg.AttestationKeys, *m.Attestation, sess.ident.UUID, now); ok {
					sess.peer.Attested = a.Network + "/" + a.Subserver
				} else {
					log.Debug("ignored invalid scope attestation")
				}
			}
		}
		s.mu.Unlock()
	case *protocol.Pos:
		if now.Sub(sess.lastPos) < 90*time.Millisecond {
			return false // above position_hz_max, dropped silently
		}
		sess.lastPos = now
		s.mu.Lock()
		if *m.Epoch == sess.peer.Epoch && sess.peer.InWorld {
			sess.peer.HasPos, sess.peer.X, sess.peer.Y, sess.peer.Z = true, *m.X, *m.Y, *m.Z
			sess.peer.PosAtMs = now.UnixMilli()
		}
		s.mu.Unlock()
		if s.cfg.LogPositions {
			log.Debug("position", "epoch", *m.Epoch, "x", *m.X, "y", *m.Y, "z", *m.Z)
		}
	case *protocol.Peers:
		s.mu.Lock()
		if *m.Epoch == sess.peer.Epoch {
			v := make(map[string]struct{}, len(m.Full))
			for _, u := range m.Full {
				if u != sess.ident.UUID {
					v[u] = struct{}{}
				}
			}
			sess.peer.Visible, sess.peersRev = v, *m.Rev
		}
		s.mu.Unlock()
	case *protocol.PeersDelta:
		s.mu.Lock()
		if *m.Epoch != sess.peer.Epoch {
			s.mu.Unlock()
			return false
		}
		if *m.Base != sess.peersRev || *m.Rev <= *m.Base {
			ep := sess.peer.Epoch
			s.mu.Unlock()
			sess.send(marshal(protocol.PeersResync{Type: "peers_resync", Epoch: ep}))
			return false
		}
		for _, u := range m.Remove {
			delete(sess.peer.Visible, u)
		}
		for _, u := range m.Add {
			if u != sess.ident.UUID && len(sess.peer.Visible) < protocol.MaxPeerList {
				sess.peer.Visible[u] = struct{}{}
			}
		}
		sess.peersRev = *m.Rev
		s.mu.Unlock()
	case *protocol.State:
		s.mu.Lock()
		sess.selfMuted = *m.Muted
		sess.peer.Muted = sess.selfMuted || sess.adminMute
		sess.peer.Deafened = *m.Deafened
		s.mu.Unlock()
	case *protocol.Ping:
		sess.send(marshal(protocol.Pong{Type: "pong", Nonce: *m.Nonce, ServerTime: now.UnixMilli()}))
	case *protocol.Bye:
		return true
	}
	return false
}

// NetworkID computes the spec 6.3 network id for tests and tools.
func NetworkID(host string, port int) string {
	h := sha256.Sum256([]byte(strings.ToLower(host) + ":" + itoa(port)))
	return "n1:" + hex.EncodeToString(h[:])[:32]
}
